latch_cli/services/cp.py

```diff
@@ -192,16 +212,46 @@ def _upload_file_chunk(
         resp = tinyrequests.request("PUT", url, data=payload)
         etag = resp.headers["ETag"]
 
-    with IO_LOCK:
-        progress_bar.update(len(payload))
+    if progress_bar is not None:
+        with IO_LOCK:
+            progress_bar.update(len(payload))
 
     return {"ETag": etag, "PartNumber": part_index + 1}
 
 
+# todo(ayush): switch this over to nucleus-data so its not this dumb
+class DownloadDirResponseData(TypedDict):
+    dir: Literal[True]
+    url: Dict[str, "DownloadResponseData"]
+
+
+class DownloadFileResponseData(TypedDict):
+    dir: Literal[False]
+    url: str
+
+
+DownloadResponseData: TypeAlias = Union[
+    DownloadDirResponseData, DownloadFileResponseData
+]
+
+
+def _get_total_download_size(response_data: DownloadResponseData) -> int:
+    if response_data["dir"]:
+        total = 0
+        for sub_res in response_data["url"].values():
+            total += _get_total_download_size(sub_res)
+        return total
+
+    with tinyrequests.get(response_data["url"], stream=True) as res:
```

Comment:
use `.head` if you just need the headers

Reply:
when i use a head request it doesn't return a value for the content-length header

Reply:
HEAD returned headers
```
x-amz-request-id: EHG541FPYDM15D9V
x-amz-id-2: oWSEmMNr7FsgrgnA4m+9YDnhsXk9KQDYauhOtE+ShKT616BFjipHSrJF5WIMCs1TvykzRy2rAWY=
Content-Type: application/xml
Date: Fri, 21 Apr 2023 22:17:55 GMT
Server: AmazonS3
```

GET returned headers
```
x-amz-id-2: NE95CH8GBZsRT/fplPC40bqBtWbNi8282qzY2wnQg2jo+KUD4Yj53/U+l3Nq8A3z9bjI2g49vRk=
x-amz-request-id: SC8ZBSGWGZPFV1FK
Date: Fri, 21 Apr 2023 22:18:12 GMT
Last-Modified: Sat, 15 Apr 2023 20:19:04 GMT
ETag: "924ec3593e423258eeb89932c8030a65-1"
x-amz-storage-class: INTELLIGENT_TIERING
x-amz-server-side-encryption: AES256
x-amz-version-id: RZ_5oFRoMHWr3Dxh2N9E5ILPhKqVpxXP
Accept-Ranges: bytes
Content-Type: text/plain
Server: AmazonS3
Content-Length: 155487310
```